pkg/authorization/authorizer/authorizer.go

```diff
@@ -45,74 +53,42 @@ func (a *openshiftAuthorizer) Authorize(attributes authorizer.Attributes) (bool,
-			// can be made on the basis of those rules that are found.
-			errs = append(errs, err)
-			continue
-		}
-
-		for _, rule := range role.Rules() {
-			matches, err := RuleMatches(attributes, rule)
-			if err != nil {
-				errs = append(errs, err)
-				continue
+	namespace := attributes.GetNamespace()
+	subjects, err := a.delegate.AllowedSubjects(attributes)
+	for _, subject := range subjects {
+		switch subject.Kind {
+		case rbac.UserKind:
+			users.Insert(subject.Name)
+		case rbac.GroupKind:
+			groups.Insert(subject.Name)
+		case rbac.ServiceAccountKind:
+			// default the namespace to namespace we're working in if
+			// it's available. This allows rolebindings that reference
+			// SAs in the local namespace to avoid having to qualify
+			// them.
+			ns := namespace
+			if len(subject.Namespace) > 0 {
+				ns = subject.Namespace
 			}
-
-			if matches {
-				users.Insert(roleBinding.Users().List()...)
-				groups.Insert(roleBinding.Groups().List()...)
+			if len(ns) >= 0 {
+				name := serviceaccount.MakeUsername(ns, subject.Name)
+				users.Insert(name)
 			}
+		default:
+			continue // TODO, should this add errs?
 		}
 	}
-
-	return users, groups, kerrors.NewAggregate(errs)
+	return users, groups, err
 }
 
-// authorizeWithNamespaceRules returns isAllowed, reason, and error.  If an error is returned, isAllowed and reason are still valid.  This seems strange
-// but errors are not always fatal to the authorization process.  It is entirely possible to get an error and be able to continue determine authorization
-// status in spite of it.  This is most common when a bound role is missing, but enough roles are still present and bound to authorize the request.
-func (a *openshiftAuthorizer) authorizeWithNamespaceRules(attributes authorizer.Attributes) (bool, string, error) {
-	allRules, ruleRetrievalError := a.ruleResolver.RulesFor(attributes.GetUser(), attributes.GetNamespace())
-
-	var errs []error
-	for _, rule := range allRules {
-		matches, err := RuleMatches(attributes, rule)
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
-		if matches {
-			if len(attributes.GetNamespace()) == 0 {
-				return true, "allowed by cluster rule", nil
-			}
-			// not 100% accurate, because the rule may have been provided by a cluster rule. we no longer have
-			// this distinction upstream in practice.
-			return true, "allowed by rule in " + attributes.GetNamespace(), nil
-		}
-	}
-	if len(errs) == 0 {
-		return false, "", ruleRetrievalError
-	}
-	if ruleRetrievalError != nil {
-		errs = append(errs, ruleRetrievalError)
+func reason(attributes authorizer.Attributes) string {
+	if len(attributes.GetNamespace()) == 0 {
+		return "allowed by cluster rule"
 	}
-	return false, "", kerrors.NewAggregate(errs)
+	// not 100% accurate, because the rule may have been provided by a cluster rule. we no longer have
+	// this distinction upstream in practice.
+	return "allowed by rule in " + attributes.GetNamespace()
```

Comment:
This isn't true, just return "allowed by openshift authorizer" until we remove it.